latch_cli/services/cp/download.py

```diff
@@ -32,44 +49,75 @@ def download(
     normalized = normalize_path(src)
     node_data = get_node_data(normalized)
 
-    node_has_children = node_data.type in {
+    can_have_children = node_data.type in {
         LDataNodeType.account_root,
         LDataNodeType.dir,
         LDataNodeType.mount,
     }
 
-    if node_has_children:
+    if can_have_children:
         endpoint = latch_config.api.data.get_signed_urls_recursive
     else:
         endpoint = latch_config.api.data.get_signed_url
 
     res = tinyrequests.post(
         endpoint,
-        headers=get_auth_header(),
+        headers={"Authorization": get_auth_header()},
         json={"path": normalized},
     )
 
     if res.status_code != 200:
-        raise ValueError(f"Failed to download {src}: {res.json()['error']}")
+        raise ValueError(
+            f"failed to fetch presigned url(s) for path {src} with code"
+            f" {res.status_code}: {res.json()['error']}"
+        )
 
     json_data = res.json()
-    if node_has_children:
+    if can_have_children:
         dir_data: GetSignedUrlsRecursiveData = json_data["data"]
 
-        if dest.exists() and not dest.is_dir():
-            raise ValueError(f"{str(dest)} is not a directory")
-
         if dest.exists() and not normalized.endswith("/"):
             dest = dest / node_data.name
 
-        urls: List[str] = []
-        dests: List[Path] = []
+        try:
+            dest.mkdir(exist_ok=True)
+        except FileNotFoundError as e:
+            raise ValueError(f"No such download destination {dest}") from e
+        except FileExistsError as e:
+            raise ValueError(f"Download destination {dest} is not a directory") from e
 
-        for rel_path, url in dir_data["urls"].items():
-            urls.append(url)
-            dests.append(dest / rel_path)
+        unconfirmed_jobs: List[DownloadJob] = []
+        confirmed_jobs: List[DownloadJob] = []
+        rejected_jobs: Set[DownloadJob] = set()
 
-        num_files = len(urls)
+        for rel_path, url in dir_data["urls"].items():
+            unconfirmed_jobs.append(DownloadJob(url, dest / rel_path))
+
+        for job in unconfirmed_jobs:
+            reject_job = False
+            for rejected in rejected_jobs:
+                if rejected.dest.parent in job.dest.parents:
+                    reject_job = True
+                    break
+
+            if reject_job:
```

Comment:
add to rejected?

Reply:
can u store strings instead in the set? so then it's just `any(x in rejected for x in job.dest.parents)`?

otherwise do `any(x in job.dest.parents for x in rejected)` which is `O(nm)` instead of `O(n)`

Reply:
no checking if a job should be rejected is `O(|rejected_jobs| * |dest.parents|)` so i want to keep `|rejected_jobs|` as small as possible and only add to it if it is the root of a new rejected tree